universal-application-tool-0.0.1/app/services/cloud/azure/BlobStorage.java

```diff
@@ -0,0 +1,241 @@
+      OffsetDateTime keyExpiration = userDelegationKey.getSignedExpiry();
+      if (userDelegationKey == null || keyExpiration.isBefore(tokenExpiration)) {
+        userDelegationKey =
+            blobServiceClient.getUserDelegationKey(
+                OffsetDateTime.now(zoneId).minus(Duration.ofMinutes(5)),
+                OffsetDateTime.now(zoneId).plus(AZURE_USER_DELEGATION_KEY_DURATION));
+      }
+      return userDelegationKey;
+    }
+
+    @Override
+    public String getSasToken(String fileName) {
+      BlobClient blobClient =
+          blobServiceClient.getBlobContainerClient(container).getBlobClient(fileName);
+
+      BlobSasPermission blobSasPermission =
+          new BlobSasPermission().setReadPermission(true).setWritePermission(true);
+
+      BlobServiceSasSignatureValues signatureValues =
+          new BlobServiceSasSignatureValues(
+                  OffsetDateTime.now(zoneId).plus(AZURE_SAS_TOKEN_DURATION), blobSasPermission)
+              .setProtocol(SasProtocol.HTTPS_ONLY);
+
+      return blobClient.generateUserDelegationSas(signatureValues, getUserDelegationKey());
+    }
+
+    @Override
+    public String getBlobUrl(String fileName) {
+      BlobClient blobClient =
+          blobServiceClient.getBlobContainerClient(container).getBlobClient(fileName);
+      return blobClient.getBlobUrl();
+    }
+  }
+
+  class AzuriteClient implements Client {
+    // Using Account SAS because Azurite emulator does not support User Delegation SAS yet
+    // See https://github.com/Azure/Azurite/issues/656
+
+    private static final String AZURE_LOCAL_CONNECTION_STRING_CONF_PATH = "azure.local.connection";
+
+    private final String connectionString;
+    private final BlobServiceClient blobServiceClient;
+    private final BlobContainerClient blobContainerClient;
+
+    AzuriteClient(Config config) {
+      this.connectionString =
+          checkNotNull(config).getString(AZURE_LOCAL_CONNECTION_STRING_CONF_PATH);
+      this.blobServiceClient =
+          new BlobServiceClientBuilder().connectionString(connectionString).buildClient();
+
+      setCorsRules(blobServiceClient);
+      this.blobContainerClient = blobServiceClient.getBlobContainerClient(container);
+      if (!blobContainerClient.exists()) {
+        blobContainerClient.create();
+      }
+    }
+
+    private void setCorsRules(BlobServiceClient blobServiceClient) {
+      BlobServiceProperties properties =
+          new BlobServiceProperties()
+              .setCors(
+                  List.of(
+                      new BlobCorsRule()
+                          // TODO: Make these more specific
+                          .setAllowedOrigins("*")
+                          .setAllowedHeaders("*")
+                          .setExposedHeaders("*")
+                          .setAllowedMethods("GET,PUT,OPTIONS")
+                          .setMaxAgeInSeconds(500)));
```

Comment:
question (blocking): will this only be used in dev? if so the cors policy isn't a big deal

Reply:
Yep, this is just for dev so I removed that TODO.